### PR TITLE
d/ec2_transit_gateway_dx_gateway_attachement - filter support

### PR DESCRIPTION
**Pull request by DrFaust92**
…source

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #11994

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
data_source_aws_ec2_transit_gateway_dx_gateway_attachement  - add filter support

```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_'
Cant run these for some reason i get error on delete of the transit gateway:

2020/03/24 23:21:59 [ERROR] <root>: eval: *terraform.EvalApplyPost, err: error deleting EC2 Transit Gateway: IncorrectState: tgw-031b5b37743c10646 has non-deleted DirectConnect Gateway Attachments: tgw-attach-0f0dc47327d18b997.


this error is consistent for me, fails regardless of the change
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
My acceptance testing:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_ -timeout 120m
=== RUN   TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_TransitGatewayIdAndDxGatewayId
=== PAUSE TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_TransitGatewayIdAndDxGatewayId
=== RUN   TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter
=== PAUSE TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter
=== CONT  TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_TransitGatewayIdAndDxGatewayId
=== CONT  TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter
--- PASS: TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_TransitGatewayIdAndDxGatewayId (1131.43s)
--- FAIL: TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter (1131.90s)
    testing.go:654: Step 0 error: errors during apply:
        
        Error: error reading EC2 Transit Gateway Direct Connect Gateway Attachment: no results found
        
          on /tmp/tf-test987078030/main.tf line 23:
          (source code not available)
        
        
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1131.949s
FAIL
GNUmakefile:26: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

I'll investigate the failure.

**Comment by ewbankkit** [comment]
Still get the same error if I simplify the relevant code to

```go
	input := &ec2.DescribeTransitGatewayAttachmentsInput{}

	input.Filters = buildEC2AttributeFilterList(map[string]string{
		"resource-type":      ec2.TransitGatewayAttachmentResourceTypeDirectConnectGateway,
		"resource-id":        d.Get("dx_gateway_id").(string),
		"transit-gateway-id": d.Get("transit_gateway_id").(string),
	})
	input.Filters = append(input.Filters, ec2TagFiltersFromMap(d.Get("tags").(map[string]interface{}))...)
	input.Filters = append(input.Filters, buildAwsDataSourceFilters(d.Get("filter").(*schema.Set))...)
	if len(input.Filters) == 0 {
		// Don't send an empty filters list; the EC2 API won't accept it.
		input.Filters = nil
	}
```

continuing investigation...

**Comment by ewbankkit** [comment]
It's a dependency ordering in the test. Change

```hcl
  filter {
    name   = "resource-id"
    values = ["${aws_dx_gateway.test.id}"]
  }
```

to

```hcl
  filter {
    name   = "resource-id"
    values = ["${aws_dx_gateway_association.test.dx_gateway_id}"]
  }
```

and the test passes:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter -timeout 120m
=== RUN   TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter
=== PAUSE TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter
=== CONT  TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter
--- PASS: TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter (1153.46s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1153.500s
```

as the data source isn't created until the attachment has been created.

**Comment by DrFaust92** [comment]
Thanks so much @ewbankkit !!!

Ive changed this accordingly

File: aws/data_source_aws_ec2_transit_gateway_dx_gateway_attachment.go [hunk 1]
```diff
@@ -18,35 +18,56 @@ func dataSourceAwsEc2TransitGatewayDxGatewayAttachment() *schema.Resource {
...
+				Values: []*string{aws.String(ec2.TransitGatewayAttachmentResourceTypeDirectConnectGateway)},
 			},
 			{
 				Name:   aws.String("transit-gateway-id"),
-				Values: []*string{aws.String(d.Get("transit_gateway_id").(string))},
+				Values: []*string{aws.String(transitGatewayId.(string))},
 			},
-		},
+		}
 	}
```
Comment (bflad) [inline aws/data_source_aws_ec2_transit_gateway_dx_gateway_attachment.go]: Same here as the other PR. 👍 (removing `resource-type` so its not duplicated in the request)

```go
	if dxGatewayIdOk {
		input.Filters = append(input.Filters, &ec2.Filter{
			Name:   aws.String("resource-id"),
			Values: []*string{aws.String(dxGatewayId.(string))},
		})
	}

	if transitGatewayIdOk {
		input.Filters = append(input.Filters, &ec2.Filter{
			Name:   aws.String("transit-gateway-id"),
			Values: []*string{aws.String(transitGatewayId.(string))},
		})
	}
```

Comment (DrFaust92) [reply aws/data_source_aws_ec2_transit_gateway_dx_gateway_attachment.go]: done + small changes to naming. 

--- PASS: TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_TransitGatewayIdAndDxGatewayId (1188.58s)
--- PASS: TestAccAWSEc2TransitGatewayDxGatewayAttachmentDataSource_filter (1256.65s)

**Pull request merged**